ICU-21650 Modified DateTimePatternGenerator to handle skeletons with …

richgillam opened the pull request:
…"e" and "c" better (allowing support for numeric day-of-week abbreviations).

<!--
Thank you for your pull request!

Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license agreement (CLA) before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html
-->

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21650
- [x] Required: The PR title must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

FrankYFTang commented on icu4c/source/i18n/dtptngen.cpp:
```diff
@@ -1648,7 +1648,7 @@ DateTimePatternGenerator::adjustFieldTypes(const UnicodeString& pattern,
                          (typeValue==UDATPG_SECOND_FIELD && (options & UDATPG_MATCH_SECOND_FIELD_LENGTH)==0) ) {
                          adjFieldLen = field.length();
-                    } else if (specifiedSkeleton) {
+                    } else if (specifiedSkeleton && reqFieldChar != LOW_C && reqFieldChar != LOW_E) {
```
Please modify the comments between line 1618 - 1640 to reflect the behavior after your change.

richgillam replied:
That seems like the wrong place to document this change; all the text there has to do with the hour characters, and this change only affects the day-of-week characters.  I added a new comment right above my change.

FrankYFTang replied:
I somehow didn't see the " a new comment right above my change." in this PR.

richgillam replied:
You're right-- it looks like I forgot to push.  I've pushed the new fix now.  Sorry about that...

FrankYFTang commented on icu4c/source/i18n/dtptngen.cpp:
```diff
@@ -1663,6 +1663,9 @@ DateTimePatternGenerator::adjustFieldTypes(const UnicodeString& pattern,
                             && (typeValue!= UDATPG_YEAR_FIELD || reqFieldChar==CAP_Y))
                             ? reqFieldChar
                             : field.charAt(0);
+                    if (c == CAP_E && adjFieldLen < 3) {
```
Why is E special here. Is there text in UTS35 mention that? If so, could you add an comment to point to that section in UTS35? Thanks

richgillam replied:
See the lines for "E", "e", and "c" in https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table. 

The description for "e" says "Local day of week number/name, format style. **Same as E except adds a numeric value that will depend on the local starting day of the week.** For this example, Monday is the first day of the week."  The description of the individual groups of letters of different lengths pretty clearly indicate that "c" is intended to work the same as "e".  For "E", on the other hand, "EEE" is the shortest supported length, and "E" is used extensively in the locale data as an alias for "EEE".

I've added a reference to UTS 35 in a comment.

FrankYFTang replied:
Could you put what you wrote above into a comment in the code here? I didn't see "added a reference to UTS 35 in a comment." in this PR yet somehow.

FrankYFTang commented on icu4c/source/test/intltest/dtptngts.cpp:
```diff
@@ -1557,4 +1558,67 @@ void IntlTestDateTimePatternGeneratorAPI::test_jConsistencyOddLocales() { // ICU
+    // generic test for DateTimePatternGenerator::getBestPattern() that can be used to test multiple
+    // bugs in the resource data
+    static const char* testCases[] = {
+        // ICU-21650: if the user asks for E, the minimum field length is 3, but if he asks for c or e, it's 1
```
Why is so? couly you put a pointer to the UTS35 which mention such here or some text to explain your rational in the comment. It is not straightforward to conclude that from glazing the UTS35. It will be nice if you can put down your logic here so others can follow. Thanks

richgillam replied:
See above.  I added a comment to the test that points to UTS 35.

FrankYFTang replied:
somehow I didn't see your comment. Did you forget to push?

FrankYFTang commented on icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/DateTimeGeneratorTest.java:
```diff
@@ -1793,4 +1793,59 @@ public void test_jConsistencyOddLocales() { // ICU-20590
+        // generic test for DateTimePatternGenerator::getBestPattern() that can be used to test multiple
+        // bugs in the resource data
+        String[] testCases = {
+            // ICU-21650: if the user asks for E, the minimum field length is 3, but if he asks for c or e, it's 1
```
same comment as C++, please add more comment to explain. Thanks

richgillam replied:
See above.  I added a comment to the test that points to UTS 35.

FrankYFTang commented on icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java:
```diff
@@ -2200,7 +2200,7 @@ private String adjustFieldTypes(PatternWithMatcher patternWithMatcher, DateTimeM
                             (type == SECOND && (options & MATCH_SECOND_FIELD_LENGTH)==0) ) {
                         adjFieldLen = fieldBuilder.length();
-                    } else if (matcherWithSkeleton != null) {
+                    } else if (matcherWithSkeleton != null && reqFieldChar != 'c' && reqFieldChar != 'e') {
```
ditto

richgillam replied:
Ditto

FrankYFTang commented on icu4j/main/classes/core/src/com/ibm/icu/text/DateTimePatternGenerator.java:
```diff
@@ -2217,6 +2217,9 @@ private String adjustFieldTypes(PatternWithMatcher patternWithMatcher, DateTimeM
                             && (type != YEAR || reqFieldChar=='Y'))
                             ? reqFieldChar
                             : fieldBuilder.charAt(0);
+                    if (c == 'E' && adjFieldLen < 3) {
```
ditto

richgillam replied:
Ditto

FrankYFTang replied:
could you add the comments in the code, not just the test code. thanks

richgillam commented:
@FrankYFTang , you were right; I forgot to push my last set of changes.  I've done so now.  Please take another look and let me know if this addresses your comments.

FrankYFTang reviewed (approved):
LGTM thanks for working on this
